File: src/main/java/xyz/chunkstories/api/net/packets/PacketInventoryPartialUpdate.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.net.packets;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import xyz.chunkstories.api.exceptions.NullItemException;
import xyz.chunkstories.api.exceptions.PacketProcessingException;
import xyz.chunkstories.api.exceptions.UndefinedItemTypeException;
import xyz.chunkstories.api.item.inventory.Inventory;
import xyz.chunkstories.api.item.inventory.InventoryTranslator;
import xyz.chunkstories.api.item.inventory.ItemPile;
import xyz.chunkstories.api.net.PacketDestinator;
import xyz.chunkstories.api.net.PacketSender;
import xyz.chunkstories.api.net.PacketSendingContext;
import xyz.chunkstories.api.net.PacketWorld;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.api.net.PacketReceptionContext;

import javax.annotation.Nullable;

public class PacketInventoryPartialUpdate extends PacketWorld {
	@Nullable
	private Inventory inventory;
	private int slotx, sloty;
	@Nullable
	private ItemPile itemPile;

	public PacketInventoryPartialUpdate(World world) {
		super(world);
	}

	public PacketInventoryPartialUpdate(World world, @Nullable Inventory inventory, int slotx, int sloty, @Nullable ItemPile newItemPile) {
		super(world);
		this.inventory = inventory;
		this.slotx = slotx;
		this.sloty = sloty;
		this.itemPile = newItemPile;
	}

	@Override
	public void send(PacketDestinator destinator, DataOutputStream out, PacketSendingContext context) throws IOException {
		InventoryTranslator.writeInventoryHandle(out, inventory);

		out.writeInt(slotx);
		out.writeInt(sloty);

		if (itemPile == null)
			out.writeInt(0);
		else {
			itemPile.saveIntoStream(world.getContentTranslator(), out);
		}
	}

	@Override
	public void process(PacketSender sender, DataInputStream in, PacketReceptionContext processor) throws IOException, PacketProcessingException {
		inventory = InventoryTranslator.obtainInventoryHandle(in, processor);

		int slotx = in.readInt();
		int sloty = in.readInt();

		try {
			itemPile = ItemPile.obtainItemPileFromStream(processor.getWorld().getContentTranslator(), in);
		} catch (NullItemException e) {
			// This is sane behavior !
			itemPile = null;
		} catch (UndefinedItemTypeException e) {
			// This is slightly more problematic.
			processor.logger().error("Undefined item: ", e);
		}

		if (inventory != null)
			inventory.setItemPileAt(slotx, sloty, itemPile);
	}

}

File: src/main/java/xyz/chunkstories/api/net/packets/PacketInventoryPartialUpdate.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.net.packets

import java.io.DataInputStream
import java.io.DataOutputStream

import xyz.chunkstories.api.exceptions.NullItemException
import xyz.chunkstories.api.exceptions.UndefinedItemTypeException
import xyz.chunkstories.api.item.inventory.Inventory
import xyz.chunkstories.api.item.inventory.*
import xyz.chunkstories.api.item.inventory.ItemPile
import xyz.chunkstories.api.net.PacketWorld
import xyz.chunkstories.api.world.World

import xyz.chunkstories.api.item.inventory.obtainItemPileFromStream
import xyz.chunkstories.api.item.inventory.saveIntoStream
import xyz.chunkstories.api.player.Player

class PacketInventoryPartialUpdate : PacketWorld {
    private var inventory: Inventory? = null

    private var x: Int = 0
    private var y: Int = 0

    private var itemPile: ItemPile? = null

    constructor(world: World) : super(world)
    constructor(world: World, inventory: Inventory?, x: Int, y: Int, newItemPile: ItemPile?) : super(world) {
        this.inventory = inventory
        this.x = x
        this.y = y
        this.itemPile = newItemPile
    }

    override fun send(dos: DataOutputStream) {
        writeInventoryHandle(dos, inventory)

        dos.writeInt(x)
        dos.writeInt(y)

        if (itemPile == null)
            dos.writeInt(0)
        else {
            itemPile!!.saveIntoStream(world.gameInstance.contentTranslator, dos)
        }
    }

    override fun receive(dis: DataInputStream, player: Player?) {
        inventory = obtainInventoryByHandle(dis, world)

        val x = dis.readInt()
        val y = dis.readInt()

        val (item, amount) =
        try {
            obtainItemPileFromStream(world.gameInstance.contentTranslator, dis)
        } catch (e: NullItemException) {
            // This is sane behavior !
            Pair(null, null)
        } catch (e: UndefinedItemTypeException) {
            // This is slightly more problematic.
            world.logger.error("Undefined item: ", e)
            throw e
        }

        if (inventory != null)
            inventory!!.setItemAt(x, y, item, amount ?: 0, true)
    }

}
